homeassistant/components/google_mail/config_flow.py

```diff
@@ -0,0 +1,79 @@
+"""Config flow for Google Mail integration."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+import logging
+from typing import Any
+
+from google.oauth2.credentials import Credentials
+from googleapiclient.discovery import build
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import config_entry_oauth2_flow
+
+from .const import DEFAULT_ACCESS, DOMAIN
+
+
+class OAuth2FlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=DOMAIN
+):
+    """Config flow to handle Google Mail OAuth2 authentication."""
+
+    DOMAIN = DOMAIN
+
+    reauth_entry: ConfigEntry | None = None
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    @property
+    def extra_authorize_data(self) -> dict[str, Any]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {
+            "scope": " ".join(DEFAULT_ACCESS),
+            # Add params to ensure we get back a refresh token
+            "access_type": "offline",
+            "prompt": "consent",
+        }
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Perform reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm reauth dialog."""
+        if user_input is None:
+            return self.async_show_form(step_id="reauth_confirm")
+        return await self.async_step_user()
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> FlowResult:
+        """Create an entry for the flow, or update existing entry."""
+        if self.reauth_entry:
+            self.hass.config_entries.async_update_entry(self.reauth_entry, data=data)
+            await self.hass.config_entries.async_reload(self.reauth_entry.entry_id)
+            return self.async_abort(reason="reauth_successful")
+
+        credentials = Credentials(data[CONF_TOKEN][CONF_ACCESS_TOKEN])
+
+        def _get_profile() -> dict[str, Any]:
+            """Get profile from inside the executor."""
+            users = build(  # pylint: disable=no-member
+                "gmail", "v1", credentials=credentials
+            ).users()
+            return users.getProfile(userId="me").execute()
+
+        email = (await self.hass.async_add_executor_job(_get_profile))["emailAddress"]
+
+        await self.async_set_unique_id(email)
+        self._abort_if_unique_id_configured()
```

Comment:
Does this integration need to ensure only one can be registered? I don't see a way that it can support multiple with the notify platform (my suggestion, i realize)

Reply:
Who doesn't have multiple Gmail accounts? lmao

I have a personal and a work one well at least the one work one that I actually read (3 if you count all the domains). Each entry creates a new notify service that the user can choose from with the email address as the name.

Reply:
Yes, agreed, having that limitation would not be good :) just wasn't sure how it works.

How are the separate services managed / identified? I am not super familiar with the how this setup works and how these are differentiated, and i don't think there is test coverage for multiple config entries so I just wanted to be certain.

Reply:
The `auth` object gets passed to the `notify` service which abstracts all that away. I've just tested this with both accounts and it is working. Perhaps a future PR can be made to ensure this going forward.

Reply:
Great, yes, as long as it works, not looking for more work here. Does it create two services with different names or something? Just curious.

Reply:
Yep, matching the email address of the account.